### ISSUE TITLE
Footnote appearing on page before marker

### ISSUE THREAD

**Issue by alerque** [simoncozens/sile#264]
Somewhere along the line (I don't know yet when this started happening or have a MWE for it), some footnotes started getting inserted on the page before their respective markers. Note footnote 31 here appears in the text on the page after its footnote.

![selection_159](https://cloud.githubusercontent.com/assets/173595/13043510/40b3d6e4-d3d1-11e5-80fa-3fc2f06505e7.png)

**Comment by simoncozens** [comment]
Looking into this now - `bug-214` is showing it.

**Comment by simoncozens** [comment]
OK, I am starting to understand this. (Fixing it is a different matter.) It's a catastrophic case of footnote splitting. The footnote is split and displayed, but then for some reason there isn't enough room on the (now reduced) page for the reference any more.

**Comment by alerque** [comment]
At first blush this works for the included test but gets stuck in some infinite loop when I tried to feed a book through it. It gets through page 6 and then just sits there with lua eating through 100% CPU and eventually most of 32G of ram.

**Comment by simoncozens** [comment]
Here's a hideously simple MWE for the infinite loop:

```TeX
\begin[papersize=432pt x 648pt,class=book]{document}
\script[src=packages/linespacing]
\set[parameter=linespacing.method,value=fit-font]
\set[parameter=linespacing.fit-font.extra-space,value=15pt plus 0.5pt minus 0.5pt]
x\skip[height=420pt].\footnote{Matta 24:13}
\end{document}
```

**Comment by alerque** [comment]
Commit fd4ee10 does not fix the infinite loop in my document (although it does for your MWE).

**Comment by simoncozens** [comment]
Are you sure? I just managed to process your document (the 8vo version) fine. The only changes between my repo and yours are that `SILE.settings.set("linespacing.fit-font.extra-space", "wibble")` should become `SILE.settings.set("linespacing.fit-font.extra-space", SILE.length.parse("wibble"))`.

**Comment by alerque** [comment]
Oh the joys of debugging. [In production](https://twitter.com/JZdziarski/status/706995222862024704).

I think we're testing past each other. Yes, my build _was_ still failing, but since 19 hours ago that's changed. I just tested again just now using what I had installed (built from b13e2c3) and it failed again. I then upgraded to the latest build (built from f9f36db) and patched my library to use `SILE.length.parse()`. The result compiles.

I'm really confused about what changed [from b13e2c3 to f9f36db](https://github.com/simoncozens/sile/compare/b13e2c3...f9f36db) that fixed this, but at least the case is repeatable for me and the HEAD version is what works for me.

I see there are new issues but I'll review them separately as the footnote markers seem to match the page with the footnotes now and the infinite loop is a no-repro in HEAD.
